### ISSUE TITLE
On the sampling operation when calculating the item exposure loss

### ISSUE THREAD

**Issue by SanaMoko** [haolun-wu/Multi-Fair-Rec#1]
As for the sampling strategy used to calculate the exposure loss, I don't understand why it is done.
`    # Sample max_pos items for each user
    neg_ids_list = []
    pos_ids_list = []

    for i in range(user_size):
        if (len(train_user_list[i]) > max_pos):
            sampled_pos_ids = np.random.choice(len(train_user_list[i]), size=max_pos, replace=False)
            tmp = [train_user_list[i][j] for j in sampled_pos_ids]
            pos_ids_list.append(tmp)
        else:
            pos_ids_list.append(train_user_list[i])
        neg_ids_list.append(negsamp_vectorized_bsearch_preverif(np.array(train_user_list[i]), item_size,
                                                                n_samp=max_pos - len(pos_ids_list[i])))

    sampled_ids = np.ones((user_size, max_pos)) * item_size
    labels = np.zeros((user_size, max_pos))

    for i in range(user_size):
        sampled_ids[i][:len(pos_ids_list[i])] = np.array(pos_ids_list[i])
        sampled_ids[i][len(pos_ids_list[i]):] = neg_ids_list[i]
        labels[i][:len(pos_ids_list[i])] = 1

    sampled_ids = torch.LongTensor(sampled_ids).to(args.device)
    labels = torch.LongTensor(labels).to(args.device)`
    `                    ranks = ranker(scores[unique_u], scores_all[unique_u])

                    # print("ranks:", ranks.shape, ranks)
                    exposure = torch.pow(args.gamma, ranks)

                    prob = F.gumbel_softmax(scores[unique_u], tau=1, hard=False)
                    sys_exposure = exposure * prob

                    genre_top_mask = genre_mask[:, sampled_ids[unique_u].long()]
                    # print("genre_mask:", genre_mask.shape)
                    # print("genre_top_mask:", genre_top_mask.shape)

                    genre_exposure = torch.matmul(genre_top_mask.reshape(genre_num, -1), sys_exposure.reshape(-1, 1))
                    # print("genre_exposure:", genre_exposure.shape)
                    genre_exposure = genre_exposure / genre_exposure.sum()

                    loss['3'] = torch.abs(torch.log(1 + torch.abs(genre_exposure - target_exposure))).sum()`
